### DISCUSSION TITLE
How to see which image owns overlay directory?

### DISCUSSION THREAD

**Discussion by abitrolly** [containers/podman#24845] [Q&A]
I was grepping for some files and got hits in  `.local/share/containers/storage/overlay/74bfb1e9ed2b863f77535942c6d4f5f94131a5610d1ebbb1dee93c40c99c8a5a/diff/usr/share/` which is `podman` image cache. How to determine to which image it belongs? The hash `74bfb1e9ed2b` doesn't match any, and container list is empty.

```sh
$ podman images
REPOSITORY                         TAG         IMAGE ID      CREATED        SIZE
registry.fedoraproject.org/fedora  latest      adc0a2cc13f6  3 weeks ago    164 MB
docker.io/library/archlinux        latest      12e1b7cf5154  8 weeks ago    462 MB
docker.io/library/ubuntu           latest      59ab366372d5  2 months ago   80.6 MB
docker.io/library/python           3.12        0218518c77be  4 months ago   1.04 GB
localhost/xxx                      latest      84a430e983b9  5 months ago   1.08 kB
docker.io/library/debian           latest      c69ac6bed067  6 months ago   121 MB
gcr.io/kaniko-project/executor     latest      8e981b24e935  6 months ago   71.9 MB
docker.io/library/ubuntu           24.04       17c0145030df  6 months ago   78.7 MB
ghcr.io/opendevin/opendevin        latest      ce6e72faf570  7 months ago   6.87 GB
docker.io/olivr/copybara           latest      7d97c9395353  22 months ago  354 MB
```

**Comment by raphendyr** [comment]
I hade the same issue. I just wrote the following script for our needs.

```sh
#!/bin/sh
set -eu

if ! [ "$1" ]; then
    echo "usage: $0 /path/to/overlay/.../diff/file/in/layer"
    exit 1
fi

if ! [ -e "$1" ]; then
    echo "The path doesn't exist on disk: $1"
    exit 1
fi

prefix=$(echo "$1" | sed -E 's,^(.*/overlay/[0-9a-f]{64})/.*$,\1,')
if ! [ "$prefix" ]; then
    echo "The path doesn't contain overlay pattern: $1"
fi

if command -v podman >/dev/null; then
    cli=podman
elif command -v docker >/dev/null; then
    cli=docker
else
    echo "No container tool found, expected podman or docker"
    exit 1
fi

$cli image ls --format='{{.Repository}}:{{.Tag}}' | while read -r image; do
    if $cli image inspect --format '{{json .GraphDriver.Data}}' "$image" | grep -qsF "$prefix"; then
        echo "$image"
    fi
done
```

It assumes all images are tagged. I don't fix that now.

**Comment by jwcard** [comment]
Yes, in general you can find the hash of a container using:

`podman inspect $CONTAINER_NAME --format "{{.GraphDriver.Data.UpperDir}}"`

**Reply by abitrolly** [reply]
@jwcard the question is about images taking space, and finding leftovers that are not linked to any image. Container list is empty.

```bash
$ podman ps -a
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
$
```

**Reply by jwcard** [reply]
You are right but I was hitting a similar thing and saw this question and posted that. Of course you can replace $CONTAINER_NAME with $IMAGE_NAME and it will still work. That's essentially what the script will do. That's a nice answer but the one I gave you can use without having to copy a script over.

**Reply by abitrolly** [reply]
I see. That still need a to be wrapped in script that will do reverse lookup for the directory.

What is the difference between `UpperDir` and `LowerDir` and `WorkDir` there?
